covid_moonshot_ml/nn/loss/MSELoss.py

```diff
@@ -0,0 +1,113 @@
+        else:
+            self.loss_function = super(MSELoss, self).forward
+
+        self.loss_type = loss_type
+
+    def forward(self, input, target, in_range, uncertainty):
+        """
+        Dispatch method for calculating loss. All arguments should be passed
+        regardless of actual loss function to keep an identical signature for
+        this class. Data is passed to `self.loss_function`.
+        """
+
+        if self.loss_type is None:
+            return self.loss_function(input, target).mean()
+        elif self.loss_type == "step":
+            return self.loss_function(input, target, in_range)
+        elif self.loss_type == "uncertainty":
+            return self.loss_function(input, target, uncertainty)
+
+    def step_loss(self, input, target, in_range):
+        """
+        Step loss calculation. For `in_range` < 0, loss is returned as 0 if
+        `input` < `target`, otherwise MSE is calculated as normal. For
+        `in_range` > 0, loss is returned as 0 if `input` > `target`, otherwise
+        MSE is calculated as normal. For `in_range` == 0, MSE is calculated as
+        normal.
+
+        Parameters
+        ----------
+        input : torch.Tensor
+            Model prediction
+        target : torch.Tensor
+            Prediction target
+        in_range : torch.Tensor
+            `target`'s presence in the dynamic range of the assay. Give a value
+            of < 0 for `target` below lower bound, > 0 for `target` above upper
+            bound, and 0 or None for inside range
+
+        Returns
+        -------
+        torch.Tensor
+            Calculated loss
+        """
+        ## Calculate loss
+        loss = super(MSELoss, self).forward(input, target)
+
+        ## Calculate mask:
+        ##  1.0 - If input or data is semiquant and prediction is inside the
+        ##    assay range
+        ##  0.0 - If data is semiquant and prediction is outside the assay range
+        mask = torch.tensor(
+            [
+                1.0 if r == 0 else ((r < 0) == (t < i))
+                for i, t, r in zip(input, target, in_range)
+            ]
+        )
+        mask = mask.to(loss.device)
+
+        return (mask * loss).mean()
+
+    def uncertainty_loss(self, input, target, uncertainty):
+        """
+        Uncertainty MSE loss calculation. Loss for each sample is calculated by
+        first dividing the difference between `input` and `target` by the
+        uncertainty in the `target` measurement.
+
+        Parameters
+        ----------
+        input : torch.Tensor
+            Model prediction
+        target : torch.Tensor
+            Prediction target
+        uncertainty : torch.Tensor
+            Uncertainty in `target` measurements
+
+        Returns
+        -------
+        """
+        ## Calculate loss
+        loss = super(MSELoss, self).forward(input, target)
+
+        ## Divide by uncertainty squared
+        loss /= uncertainty**2
```

Comment:
If the [pytorch `MSELoss`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py#L532-L536) has already applied the `mean` reduction across the data, isn't it too late to scale the data by `uncertainty` if each measurement has a different value of `uncertainty`?

Reply:
In [line 19](https://github.com/choderalab/covid-moonshot-ml/blob/add-loss-functions/covid_moonshot_ml/nn/loss/MSELoss.py#L19) the `MSELoss` class is initialized with `reduction="none"`, so the squared error for each example is returned.